shared/src/balancer_event_handler.rs

```diff
@@ -54,15 +130,124 @@ impl PoolSpecialization {
+            t => Err(anyhow!("Invalid PoolSpecialization value {} (> 2)", t)),
         }
     }
 }
 
-#[derive(Default)]
-pub struct BalancerPools {}
-
 impl BalancerPools {
+    // All insertions happen in one transaction.
+    fn insert_events(&mut self, events: Vec<(EventIndex, BalancerEvent)>) -> Result<()> {
+        for (index, event) in events {
+            match event {
+                BalancerEvent::PoolRegistered(event) => self.insert_pool(index, event),
+                BalancerEvent::TokensRegistered(event) => self.insert_token_data(index, event),
+            };
+        }
+        Ok(())
+    }
+
+    fn insert_pool(&mut self, index: EventIndex, registration: PoolRegistered) {
+        match self.pools.get_mut(&registration.pool_id) {
+            None => {
+                // PoolRegistered event is first to be processed, we leave tokens empty
+                // and update when TokenRegistration event is processed.
+                tracing::debug!(
+                    "Creating Balancer Pool with id {:?} from PoolRegistration event (tokens pending...)",
+                    &registration.pool_id
+                );
+                self.pools.insert(
+                    registration.pool_id,
+                    WeightedPool::from((index, registration)),
+                );
+            }
+            Some(existing_pool) => {
+                // If exists before PoolRegistration, then only pool_id and tokens are known
+                existing_pool.update_from_event(BalancerEvent::PoolRegistered(registration));
+                tracing::debug!(
+                    "Pool Address and specialization recovered for existing Balancer pool with id {:}", existing_pool.pool_id
+                );
+            }
+        }
+    }
+
+    fn insert_token_data(&mut self, index: EventIndex, registration: TokensRegistered) {
+        let pool_id = &registration.pool_id;
+        match self.pools.get_mut(pool_id) {
+            None => {
+                // TokensRegistered event received before PoolRegistered, leaving pool_address
+                // empty until PoolRegistration event is processed.
+                tracing::debug!(
+                    "Creating Balancer Pool with id {:?} from TokensRegistration event (pool address pending...)", pool_id
+                );
+                self.pools
+                    .insert(*pool_id, WeightedPool::from((index, registration.clone())));
+            }
+            Some(existing_pool) => {
+                // If pool exists before token registration, then only pool_id and address known
+                existing_pool
+                    .update_from_event(BalancerEvent::TokensRegistered(registration.clone()));
+                tracing::debug!(
+                    "Pool Tokens recovered for existing Balancer pool with id {:?}",
+                    pool_id
+                )
+            }
+        }
+
+        // In either of the above cases we can now populate pools_by_token
```

Comment:
I see that these events happen in the same transactions, but conecptually shouldn't we only advertise pools that are "completely" initialized?

Reply:
Yes, although I think this will be covered by the fact that we will introduce `WeightedPoolBuilder` and only expose those which are built.